ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+      , (10, pure $ Action SyncLedger)
+      , (10, do
+            b  <- arbitrary
+            ls <- arbitrary `suchThat` (  (`notElem` modelSeenStates)
+                                         . castPoint
+                                         . getTip
+                                       ) -- FIXME maybe not needed?
+            pure $ Event $ ChangeLedger ls b)
+      , (10, pure $ Action GetSnapshot)
+      ]
+
+data Snap blk = Snap
+ { snapTxs         :: ![(GenTx blk, TicketNo)] }
+
+mock ::
+     Model blk Symbolic
+  -> Command blk Symbolic
+  -> GenSym (Response blk Symbolic)
+mock model = \case
+  Action (NewTxs _)      -> pure Void
+  Action SyncLedger      -> pure Void
+  Action GetSnapshot     -> pure $ GotSnapshot Snap {
+         snapTxs = modelTxs
+       }
+  Event (ChangeLedger _ _) -> pure Void
+  where
+    Model {
+        modelTxs
+      } = model
+
+transitions ::
+     ( Eq (GenTx blk)
+     , Eq (TickedLedgerState blk ValuesMK)
+     , HasTxId (GenTx blk)
+     , LedgerSupportsMempool blk
+     , ToExpr (GenTx blk)
+     , ToExpr (LedgerState blk ValuesMK)
+     , ValidateEnvelope blk
+     )
+  => LedgerConfig blk
+  -> Model blk r
+  -> Command blk r
+  -> Response blk r
+  -> Model blk r
+transitions cfg model cmd resp = case (cmd, resp) of
+  (Action (NewTxs txs), Void) ->
+    if reachable
+    then
+      let nextTicket = succ $ modelLastTicket model
+          (res, tk, _, st) = foldTxs cfg nextTicket (currentRemainingCap model)
+                              mtip
+                            $ txs
+          modelTxs'     = mtxs ++
+                        zip [ txForgetValidated vtx
+                            | MempoolTxAdded vtx <- fst res
+                            ]
+                            [nextTicket..]
+      in
+        model {
+           modelMempoolTip = (st, reachable)
+         , modelTxs        = modelTxs'
+         , modelLastTicket = if tk == zeroTicketNo
+                             then zeroTicketNo
+                             else pred tk
+         , modelCapacity   = MempoolCapacityBytes 400
+         , modelMempoolWasFilled = modelMempoolWasFilled || (not . null . snd $ res)
+         }
+    else
+      let model' = transitions cfg model (Action SyncLedger) Void
+      in transitions cfg model' cmd Void
+
+  (Event (ChangeLedger l' b'), Void) ->
+        model { modelMempoolTip  = (mtip, reachable && not b')
+              , modelLedgerDBTip = l'
+              , modelSeenStates  = Set.insert (castPoint $ getTip l') seenStates
+              }
+  (Action GetSnapshot, GotSnapshot{}) -> model
+  (Action SyncLedger, Void) ->
+    if mtip == tick cfg dbtip
+    then model
+    else
+      let modelTxs'       = fmap fst mtxs
+          (res, _, _, st) = foldTxs cfg zeroTicketNo modelCapacity
+                              (tick cfg dbtip)
+                              modelTxs'
+          filteredTxSeq   =
```

Comment:
Would it make sense to factor this out or place this logic inside `foldTxs`? It seems to obscure the transition logic otherwise.